lib/location/loc_core.c

```diff
@@ -0,0 +1,501 @@
+		/* Check if the method is valid */
+		method_api = loc_method_api_validity_get(config->methods[i].method);
+		if (method_api == NULL) {
+			LOG_ERR("Location method (%d) not supported", config->methods[i].method);
+			return -EINVAL;
+		}
+		/* Validate method specific parameters */
+		if (methods_supported[i]->validate_params) {
+			err = methods_supported[i]->validate_params(&config->methods[i]);
+			if (err) {
+				return err;
+			}
+		}
+	}
+	return 0;
+}
+
+void loc_core_config_log(const struct loc_config *config)
+{
+	enum loc_method type;
+	const struct loc_method_api *method_api;
+
+	LOG_DBG("Location configuration:");
+
+	LOG_DBG("  Methods count: %d", config->methods_count);
+	LOG_DBG("  Interval: %d", config->interval);
+	LOG_DBG("  List of methods:");
+
+	for (uint8_t i = 0; i < config->methods_count; i++) {
+		type = config->methods[i].method;
+		method_api = loc_method_api_validity_get(type);
+
+		LOG_DBG("    Method #%d", i);
+		if (method_api != NULL) {
+			LOG_DBG("      Method type: %s (%d)",
+				log_strdup(method_api->method_string), type);
+		} else {
+			LOG_DBG("      Method type: unknown (%d)", type);
+		}
+
+		if (type == LOC_METHOD_GNSS) {
+			LOG_DBG("      Timeout: %d", config->methods[i].gnss.timeout);
+			LOG_DBG("      Accuracy: %s (%d)",
+				log_strdup(loc_core_gnss_accuracy_str(
+					config->methods[i].gnss.accuracy)),
+				config->methods[i].gnss.accuracy);
+			LOG_DBG("      Number of fixes: %d",
+				config->methods[i].gnss.num_consecutive_fixes);
+		} else if (type == LOC_METHOD_CELLULAR) {
+			LOG_DBG("      Timeout: %d", config->methods[i].cellular.timeout);
+		} else if (type == LOC_METHOD_WIFI) {
+			LOG_DBG("      Timeout: %d", config->methods[i].wifi.timeout);
+		}
+	}
+}
+
+static void current_loc_config_set(const struct loc_config *config)
+{
+	assert(config->methods_count <= LOC_MAX_METHODS);
+
+	if (config == &current_loc_config) {
+		/* This is hit in periodic mode */
+		return;
+	}
+
+	current_loc_config.methods_count = config->methods_count;
+	current_loc_config.interval = config->interval;
+
+	memset(current_loc_config.methods, 0, sizeof(struct loc_method_config) * LOC_MAX_METHODS);
+	memcpy(current_loc_config.methods, config->methods,
+		sizeof(struct loc_method_config) * config->methods_count);
+}
+
+int loc_core_location_get(const struct loc_config *config)
+{
+	int err;
+	enum loc_method requested_loc_method;
+
+	current_loc_config_set(config);
+	/* Location request starts from the first method */
+	current_loc_method_index = 0;
+	requested_loc_method = config->methods[current_loc_method_index].method;
+	LOG_DBG("Requesting location with '%s' method",
+		(char *)loc_method_api_get(requested_loc_method)->method_string);
+	loc_core_current_event_data_init(requested_loc_method);
+	err = loc_method_api_get(requested_loc_method)->location_get(
+		&config->methods[current_loc_method_index]);
+
+	return err;
+}
+
+void loc_core_event_cb_error(void)
+{
+	current_event_data.id = LOC_EVT_ERROR;
```

Comment:
There's also a place for error cause in the event. Either we should add support for passing the error cause or remove it from the API.

Reply:
Tuomas will remove cause from API